src/uct/sm/scopy/base/scopy_ep.c

```diff
@@ -3,21 +3,203 @@
+    tx->rkey            = rkey;
+    tx->remote_addr     = remote_addr;
+    tx->op              = tx_op;
+    tx->consumed_length = 0;
+    tx->iov_cnt         = 0;
+    tx->total_length    = 0;
+    ucs_iov_iter_init(&tx->iov_iter);
+    for (iov_it = 0; iov_it < iov_cnt; iov_it++) {
+        if (uct_iov_get_length(&iov[iov_it]) == 0) {
+            /* Avoid zero-length IOV elements */
+            continue;
+        }
+
+        tx->iov[tx->iov_cnt] = iov[iov_it];
+        tx->total_length    += tx->iov[tx->iov_cnt].length;
+        tx->iov_cnt++;
+    }
+
+    ucs_list_head_init(&tx->flush_comp_list);
+
+    if (tx_op == UCT_SCOPY_TX_PUT_ZCOPY) {
+        UCT_TL_EP_STAT_OP(ucs_derived_of(tl_ep, uct_base_ep_t), PUT, ZCOPY,
+                          tx->total_length);
+    } else {
+        UCT_TL_EP_STAT_OP(ucs_derived_of(tl_ep, uct_base_ep_t), GET, ZCOPY,
+                          tx->total_length);
+    }
+
+    if (tx->total_length == 0) {
+        uct_scopy_trace_data(tx);
+        ucs_mpool_put_inline(tx);
+        return UCS_OK;
+    }
+
+    ep->last_tx = tx;
+    iface->outstanding++;
+    ucs_arbiter_elem_init(&tx->arb_elem);
+    ucs_arbiter_group_push_elem(&ep->arb_group, &tx->arb_elem);
+    ucs_arbiter_group_schedule(&iface->arbiter, &ep->arb_group);
+
+    return UCS_INPROGRESS;
+}
+
+ucs_status_t uct_scopy_ep_put_zcopy(uct_ep_h tl_ep, const uct_iov_t *iov,
+                                    size_t iov_cnt, uint64_t remote_addr,
+                                    uct_rkey_t rkey, uct_completion_t *comp)
+{
+    return uct_scopy_ep_prepare_tx(tl_ep, iov, iov_cnt, remote_addr,
+                                   rkey, comp, UCT_SCOPY_TX_PUT_ZCOPY);
+}
+
+ucs_status_t uct_scopy_ep_get_zcopy(uct_ep_h tl_ep, const uct_iov_t *iov,
+                                    size_t iov_cnt, uint64_t remote_addr,
+                                    uct_rkey_t rkey, uct_completion_t *comp)
+{
+    return uct_scopy_ep_prepare_tx(tl_ep, iov, iov_cnt, remote_addr,
+                                   rkey, comp, UCT_SCOPY_TX_GET_ZCOPY);
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+uct_scopy_ep_comp_tx(uct_scopy_ep_t *ep, uct_scopy_tx_t *tx,
+                     ucs_status_t status, size_t comp_size)
+{
+    uct_scopy_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                              uct_scopy_iface_t);
+    uct_scopy_comp_t *flush_comp, *tmp_flush_comp;
+
+    tx->consumed_length += comp_size;
+
+    uct_scopy_trace_data(tx);
+
+    if ((status != UCS_OK) || (tx->consumed_length == tx->total_length)) {
+        if (tx->comp != NULL) {
+            uct_invoke_completion(tx->comp, status);
+        }
+
+        ucs_list_for_each_safe(flush_comp, tmp_flush_comp,
+                               &tx->flush_comp_list, list_elem) {
+            uct_invoke_completion(flush_comp->comp, status);
+            ucs_list_del(&flush_comp->list_elem);
+            ucs_mpool_put_inline(flush_comp);
+        }
```

Comment:
can we add flush operations to the arbiter (as tx op) instead of keeping a list in each op and keeping ep->last_tx?
the arbiter will dispatch flush ops after the tx ops before are finished